main/src/library/Graph.flix

```diff
@@ -101,4 +101,68 @@ namespace Graph {
         Reachable(n1, n2) :- Reachable(n1, m), Edge(m, n2).
     }
 
+    ///
+    /// Optionally returns the shortest distance from `src` to `dst` in the weighted directed graph `g`.
+    ///
+    pub def distance(src: {src :: t}, dst: {dst :: t}, g: m[(t, d, t)]): Option[d] with Foldable[m], Boxable[t], Boxable[d], Add[d] =
+        let edges = inject g into Edge;
+        let x = query edges, dists(src.src) select x from Distance(n, x) where n == dst.dst;
+        List.minimum(x)
+
+    ///
+    /// Returns the shortest distance from `src` to every other vertex in the weighted directed graph `g`.
+    ///
+    pub def distances(src: t, g: m[(t, d, t)]): Map[t, Option[d]] with Foldable[m], Boxable[t], Boxable[d], Add[d] =
+        let edges = inject g into Edge;
+        let allDists = #{
+            // Collects all the distances `d` from `src` to `n` into a Set.
+            AllDists(n; Set#{d}) :- Distance(n, d).
+        };
+        let reachableDists = query edges, dists(src), allDists select (x, Set.minimum(s)) from AllDists(x; s) |> List.toMap;
+        let unreachableDists = unreachableWithDists(src, g)
+                                |> List.foldRight((x, acc) -> Map.insert(x, None, acc), Map.empty());
+        Map.union(reachableDists, unreachableDists)
+
+    ///
+    /// Helper function for `distance` and `distances`.
+    /// Returns a Datalog program which computes distances to all reachable nodes when given a set of `Edge` facts and a `src`.
+    ///
+    def dists(src: t): #{ Edge(t, d, t) | r } with Boxable[t], Boxable[d], Add[d] = #{
+        // Distance `d` from `src` to `n`.
+        Distance(n, d) :- Edge(src, d, n), if n != src.
```

Comment:
Does this not become a problem, when we use `d` instead of `Int32`. This means I cannot write something like:
```
def dists(src: t): #{ Node(t), Edge(t, d, t), Distance(t, d) | r } with Boxable[t], Boxable[d], Add[d] = #{
    // All nodes have a distance if 0 to itself.
    Distance(n, 0) :- Node(n).
    // If `src` can reach `m` with distance `d1` and
    // there is an edge from `m` to `n` with distance `d2`
    // then `src` can reach `n` with distance `d1 + d2`.
    Distance(n, d1 + d2) :- Distance(m, d1), Edge(m, d2, n), if n != src.
}
```

Reply:
Hmm good point. I think it should be `LowerBound.lowerbound()` which will be 0 for Int32. This should already come with the required typeclasses I believe.

Reply:
We only have `LowerBound.minValue()` which is `-2147483648` for `Int32`

Reply:
Yes, my mistake. hmmm. I still believe that you would want nodes to have a distance of 0 to themselves and that we should only find the minimum for performance. Maybe we could add a Metric type class to this namespace? It needs a `zero`, `add`, the lattice operations. I'm not sure. Can you chime in @magnus-madsen?

Reply:
I think we need to define our own lattices and "flip it upside down" like on the slides.

Reply:
So then `d` becomes `Int32` or do we just use the lowerbound and then facilitate lattice wrapping of `Int32`? bonus hack, we could reify `d` and customize what happens for `IntX`s